### PR TITLE
[TRTLLM-5273]feat/Use full attention mask if Llama3 is used as encoder and fix EarlyStopDecoder unsqueeze bug

### PR DESCRIPTION
**Pull request by nvrohanv**
# [TRTLLM-5273]feat/Use full attention mask if Llama3 is used as encoder and fix EarlyStopDecoder unsqueeze bug

## Description

This PR adds in a flag for bidirectional_attention to `modeling_llama.py`. This is to support some customer use-cases for llama-based embedding models, see Jira TRTLLM-5273 for more details.

Additionally it fixes a bug in EarlyStopDecoder where 1d logits were being unsqueezed incorrectly, leading to incompatibility with LogitsStorage. 

Lastly it hardcodes `LlamaForTextEmbedding` as a non-generation model for now so that the customer can make their own out-of-tree example with their custom model and still leverage EarlyStopDecoder. I am in discussions with Erin Ho and Freddy Qi about how we can eliminate the need to hardcode this in the future. 

**Things To Review**

1. Should this bidirectional attention be supported via flag like the current implementation or should it be a separate class?

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/models/modeling_llama.py [hunk 1]
```diff
@@ -527,6 +532,7 @@ def forward(
             position_ids=position_ids,
             hidden_states=hidden_states,
             attn_metadata=attn_metadata,
+            attention_mask=self.attention_mask,
```
Comment (symphonylyh) [inline tensorrt_llm/_torch/models/modeling_llama.py:535]: is attention_mask usually a init-stage property at LlamaAttention() or a runtime input parameter? @qixiang-99 
if init, how about merge it inside model_config and handle that in the attention layer?

File: tensorrt_llm/_torch/models/modeling_llama.py [hunk 2]
```diff
@@ -728,7 +734,9 @@ def forward(
 @support_pp
 class LlamaModel(DecoderModel):
 
-    def __init__(self, model_config: ModelConfig[LlamaConfig]):
+    def __init__(self,
+                 model_config: ModelConfig[LlamaConfig],
+                 bidirectional=False):
```
Comment (symphonylyh) [inline tensorrt_llm/_torch/models/modeling_llama.py:739]: for example, I think other people may complain about adding a dangling field here

File: tensorrt_llm/_torch/models/modeling_llama.py (hunk 1 above)
Comment (qixiang-99) [reply tensorrt_llm/_torch/models/modeling_llama.py:535]: In most cases I've encountered, the attention_mask can theoretically be determined during init-stage. However, the attention forward method mandates the attention_mask as an input ([link](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/_torch/attention_backend/interface.py#L530)).

File: tensorrt_llm/_torch/models/modeling_llama.py (hunk 2 above)
Comment (qixiang-99) [reply tensorrt_llm/_torch/models/modeling_llama.py:739]: I agree with this. The primary purpose of model_config is to maintain a clean argument list, so some refactoring might be necessary.

Comment (nvrohanv) [reply tensorrt_llm/_torch/models/modeling_llama.py:739]: Got it in that case should I make a LlamaWithBidirectionalAttention class?

File: tensorrt_llm/_torch/models/modeling_llama.py (hunk 1 above)
Comment (nvrohanv) [reply tensorrt_llm/_torch/models/modeling_llama.py:535]: Resolved by just using model_config.is_generation. For now since there aren't any requests for encoder models with custom embeddings this will do and we can revisit if and when a custom embedding case comes.

Comment (symphonylyh) [reply tensorrt_llm/_torch/models/modeling_llama.py:535]: attention_mask --> attention_mask_type you plan to do in a follow-up chore PR, do I understand it correctly?

**Review by symphonylyh — APPROVED** [review]

File: tensorrt_llm/_torch/models/modeling_llama.py (hunk 1 above)
Comment (nvrohanv) [reply tensorrt_llm/_torch/models/modeling_llama.py:532]: Yes, that is the plan

**Comment by qixiang-99** [comment]
LGTM
